File: src/main/kotlin/com/boobalan/grokking/coding/tcs/Ajju.java
package com.boobalan.grokking.coding.tcs;

import org.jetbrains.annotations.NotNull;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Ajju {

    public static void main(String[] args) {
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(System.in))) {
            String line;
            line = reader.readLine();
            String[] parts = line.split("\\s+");

            Integer[] numbers = new Integer[parts.length];
            for (int i = 0; i < parts.length; i++) {
                numbers[i] = Integer.parseInt(parts[i]);
            }
            solve(numbers);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void solve(Integer[] pillarHeights) {
        if (pillarHeights.length == 1) {
            byte var1 = 0;
            System.out.println(var1);
        }

        ArrayList<Pillar> pillarsSortedByHeight = new ArrayList<>();
        int i = 0;

        for(int var3 = pillarHeights.length; i < var3; ++i) {
            pillarsSortedByHeight.add(new Pillar(pillarHeights[i], i));
        }


        pillarsSortedByHeight.sort((o1, o2) -> o2.getHeight().compareTo(o1.getHeight()));
        Object var10000 = pillarsSortedByHeight.get(0);
        Pillar leftPillar = (Pillar)var10000;
        var10000 = pillarsSortedByHeight.get(0);
        Pillar rightPillar = (Pillar)var10000;
        int areaToFill = 0;
        int j = 1;

        while(true) {
            int var13 = pillarsSortedByHeight.size();
            if (1 > j) {
                break;
            }

            if (var13 <= j || leftPillar.getPosition() == 0 && rightPillar.getPosition() == pillarHeights.length - 1) {
                break;
            }

            var10000 = pillarsSortedByHeight.get(j++);
            Pillar pillar = (Pillar)var10000;
            if (pillar.getPosition() < leftPillar.getPosition()) {
                areaToFill += pillar.getHeight() * (leftPillar.getPosition() - pillar.getPosition() - 1);
                leftPillar = pillar;
            } else if (pillar.getPosition() > rightPillar.getPosition()) {
                areaToFill += pillar.getHeight() * (pillar.getPosition() - rightPillar.getPosition() - 1);
                rightPillar = pillar;
            } else {
                areaToFill -= pillar.getHeight();
            }
        }

        int restOfSum = 0;
        int k = j;

        for(int var8 = pillarHeights.length; k < var8; ++k) {
            restOfSum += ((Pillar)pillarsSortedByHeight.get(k)).getHeight();
        }

        areaToFill -= restOfSum;
        System.out.println(areaToFill);
    }

    static class Pillar {
        private Integer height;
        private Integer position;

        public Pillar(final Integer height, final Integer position) {
            this.height = height;
            this.position = position;
        }

        public Integer getPosition() {
            return position;
        }

        public void setPosition(final Integer position) {
            this.position = position;
        }

        public Integer getHeight() {
            return height;
        }

        public void setHeight(final Integer height) {
            this.height = height;
        }
    }

}

File: src/main/kotlin/com/boobalan/grokking/coding/tcs/Ajju.kt
package com.boobalan.grokking.coding.tcs

import java.io.BufferedReader
import java.io.InputStreamReader

fun main(args: Array<String>) {

//    val reader = BufferedReader(InputStreamReader(System.`in`))
//
//    val readLine = reader.readLine()
//
//    val pillarHeights = readLine?.split(" ")?.map { it.toInt() }?.toTypedArray()
//
//    pillarHeights?.let { solve(it) }

    solve()


}

fun solve() {
    val pillarHeights = intArrayOf(1, 0, 2, 1, 0, 6, 2, 1, 0, 2)
//    val pillarHeights = intArrayOf(0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1)


    if (pillarHeights.size == 1) {
        println(0)
    }

    val pillarsSortedByHeight: ArrayList<Pillar> = ArrayList()
    for (i in pillarHeights.indices) {
        pillarsSortedByHeight.add(Pillar(pillarHeights[i], i))
    }

    pillarsSortedByHeight.sortWith { o1, o2 -> o2.height.compareTo(o1.height) }

    var leftPillar = pillarsSortedByHeight[0]
    var rightPillar = pillarsSortedByHeight[0]
    var areaToFill = 0

    var j = 1

    while (j in 1 until pillarsSortedByHeight.size && (leftPillar.position != 0 || rightPillar.position != pillarHeights.size-1)) {
        val pillar= pillarsSortedByHeight[j++]
        when {
            pillar.position < leftPillar.position -> {
                // in left side
                areaToFill += pillar.height * (leftPillar.position - pillar.position -1)
                leftPillar = pillar;

            }
            pillar.position > rightPillar.position -> {
                // in right side
                areaToFill += pillar.height * (pillar.position - rightPillar.position -1)
                rightPillar = pillar;
            }
            else -> {
                // in range
                areaToFill -= pillar.height
            }
        }
    }

    var restOfSum = 0
    for (k in j until pillarHeights.size) {
        restOfSum += pillarsSortedByHeight[k].height
    }

    areaToFill -= restOfSum

    println(areaToFill)


}

data class Pillar(val height: Int, val position: Int)